File: src/Riftbound.Contracts/Protocol.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Riftbound.Contracts;

public static class ProtocolDefaults
{
    public const int ProtocolVersion = 1;
    public const int SchemaVersion = 1;
}

public enum MessageType
{
    JOIN,
    RECONNECT,
    START,
    READY,
    ACT,
    REACT,
    PASS,
    END_TURN,
    PROMPT,
    SNAPSHOT,
    EVENTS,
    ERROR,
    PING,
    PONG
}

public sealed record WsClientMessage(
    MessageType Type,
    string RoomId,
    string PlayerId,
    string? ClientIntentId = null,
    string? ReconnectToken = null,
    JsonElement? Cmd = null,
    int ProtocolVersion = ProtocolDefaults.ProtocolVersion,
    int SchemaVersion = ProtocolDefaults.SchemaVersion);

public sealed record WsServerMessage(
    MessageType Type,
    string RoomId,
    string PlayerId,
    long ServerTick,
    object? Payload,
    int ProtocolVersion = ProtocolDefaults.ProtocolVersion,
    int SchemaVersion = ProtocolDefaults.SchemaVersion);

public static class ErrorCodes
{
    public const string PlayerIdRequired = "PLAYER_ID_REQUIRED";
    public const string PlayerNotInRoom = "PLAYER_NOT_IN_ROOM";
    public const string RoomFull = "ROOM_FULL";
    public const string InvalidReconnectToken = "INVALID_RECONNECT_TOKEN";
    public const string ClientIntentIdRequired = "CLIENT_INTENT_ID_REQUIRED";
    public const string ClientIntentConflict = "CLIENT_INTENT_CONFLICT";
    public const string MatchNotStarted = "MATCH_NOT_STARTED";
    public const string MatchFinished = "MATCH_FINISHED";
    public const string UnsupportedCommand = "UNSUPPORTED_COMMAND";
    public const string PhaseNotAllowed = "PHASE_NOT_ALLOWED";
    public const string InsufficientCost = "INSUFFICIENT_COST";
    public const string InvalidTarget = "INVALID_TARGET";
    public const string InvalidPayload = "INVALID_PAYLOAD";
    public const string CardNotInHand = "CARD_NOT_IN_HAND";
    public const string InvalidDeck = "INVALID_DECK";
    public const string UnsupportedCardBehavior = "UNSUPPORTED_CARD_BEHAVIOR";
    public const string RecoveryInconsistent = "RECOVERY_INCONSISTENT";
    public const string PromptExpired = "PROMPT_EXPIRED";
}

public static class CommandTypes
{
    public const string Ready = "READY";
    public const string SubmitDeck = "SUBMIT_DECK";
    public const string Mulligan = "MULLIGAN";
    public const string PassPriority = "PASS_PRIORITY";
    public const string PassFocus = "PASS_FOCUS";
    public const string Pass = "PASS";
    public const string EndTurn = "END_TURN";
    public const string Surrender = "SURRENDER";
    public const string PlayCard = "PLAY_CARD";
    public const string ActivateAbility = "ACTIVATE_ABILITY";
    public const string LegendAct = "LEGEND_ACT";
    public const string HideCard = "HIDE_CARD";
    public const string TapRune = "TAP_RUNE";
    public const string RecycleRune = "RECYCLE_RUNE";
    public const string RevealCard = "REVEAL_CARD";
    public const string MoveUnit = "MOVE_UNIT";
    public const string AssembleEquipment = "ASSEMBLE_EQUIPMENT";
    public const string DeclareBattle = "DECLARE_BATTLE";
    public const string PayCost = "PAY_COST";
    public const string AssignCombatDamage = "ASSIGN_COMBAT_DAMAGE";
    public const string OrderTriggers = "ORDER_TRIGGERS";
}

public sealed record ErrorDto(
    string Code,
    string Message);

public sealed record PlayerSessionDto(
    string PlayerId,
    string Seat,
    string ReconnectToken);

public abstract record GameCommand(string CmdType);

public sealed record ReadyCommand() : GameCommand("READY");

public sealed record SubmitDeckCommand(
    string LegendCardNo,
    string ChampionCardNo,
    IReadOnlyList<string> MainDeck,
    IReadOnlyList<string> RuneDeck,
    IReadOnlyList<string> Battlefields) : GameCommand("SUBMIT_DECK");

public sealed record MulliganCommand(
    IReadOnlyList<string> HandObjectIds) : GameCommand("MULLIGAN");

public sealed record PassPriorityCommand() : GameCommand("PASS_PRIORITY");

public sealed record PassFocusCommand() : GameCommand("PASS_FOCUS");

public sealed record PassCommand() : GameCommand("PASS");

public sealed record EndTurnCommand() : GameCommand("END_TURN");

public sealed record SurrenderCommand() : GameCommand("SURRENDER");

public sealed record PlayCardCommand(
    string SourceObjectId,
    string CardNo,
    IReadOnlyList<string> TargetObjectIds,
    string Mode = "",
    IReadOnlyList<string>? OptionalCosts = null,
    string Destination = "") : GameCommand("PLAY_CARD");

public sealed record ActivateAbilityCommand(
    string SourceObjectId,
    string AbilityId,
    IReadOnlyList<string> TargetObjectIds,
    IReadOnlyList<string>? OptionalCosts = null) : GameCommand("ACTIVATE_ABILITY");

public sealed record LegendActCommand(
    string SourceObjectId,
    string AbilityId,
    IReadOnlyList<string> TargetObjectIds,
    IReadOnlyList<string>? OptionalCosts = null) : GameCommand("LEGEND_ACT");

public sealed record HideCardCommand(
    string SourceObjectId,
    string CardNo,
    string Destination = "",
    IReadOnlyList<string>? OptionalCosts = null) : GameCommand("HIDE_CARD");

public sealed record TapRuneCommand(
    string SourceObjectId) : GameCommand("TAP_RUNE");

public sealed record RecycleRuneCommand(
    string SourceObjectId) : GameCommand("RECYCLE_RUNE");

public sealed record RevealCardCommand(
    string SourceObjectId,
    string CardNo,
    IReadOnlyList<string> TargetObjectIds,
    string Mode = "",
    IReadOnlyList<string>? OptionalCosts = null,
    string Destination = "") : GameCommand("REVEAL_CARD");

public sealed record MoveUnitCommand(
    string SourceObjectId,
    string Origin = "",
    string Destination = "",
    IReadOnlyList<string>? OptionalCosts = null) : GameCommand("MOVE_UNIT");

public sealed record AssembleEquipmentCommand(
    string SourceObjectId,
    string TargetObjectId = "",
    IReadOnlyList<string>? OptionalCosts = null) : GameCommand("ASSEMBLE_EQUIPMENT");

public sealed record DeclareBattleCommand(
    string BattlefieldId = "",
    IReadOnlyList<string>? AttackerObjectIds = null,
    IReadOnlyList<string>? DefenderObjectIds = null,
    IReadOnlyList<string>? OptionalCosts = null,
    IReadOnlyList<string>? BattlefieldTargetObjectIds = null) : GameCommand("DECLARE_BATTLE");

public sealed record PayCostCommand(
    string PaymentId = "",
    string PaymentWindow = "",
    IReadOnlyList<string>? PaymentChoiceIds = null) : GameCommand(CommandTypes.PayCost);

public sealed record CombatDamageAssignmentDto(
    string SourceObjectId,
    string TargetObjectId,
    int Damage);

public sealed record AssignCombatDamageCommand(
    string BattleId = "",
    string BattlefieldId = "",
    IReadOnlyList<CombatDamageAssignmentDto>? Assignments = null) : GameCommand(CommandTypes.AssignCombatDamage);

public sealed record OrderTriggersCommand(
    IReadOnlyList<string>? TriggerIds = null,
    IReadOnlyList<string>? OrderedTriggerIds = null) : GameCommand(CommandTypes.OrderTriggers);

public sealed record UnsupportedCommand(string RawCmdType, JsonElement? Payload = null)
    : GameCommand(RawCmdType);

public sealed record PlayerIntent(
    string IntentId,
    string PlayerId,
    string CommandType);

public sealed record GameEvent(
    string Kind,
    string Description,
    IReadOnlyDictionary<string, object?> Payload);

public sealed record SnapshotDto(
    long Tick,
    int TurnNumber,
    string ActivePlayerId,
    IReadOnlyDictionary<string, object?> Players,
    IReadOnlyDictionary<string, object?> Lanes,
    IReadOnlyList<object?> Stack,
    IReadOnlyDictionary<string, object?> Timing,
    string TurnState);

public static class PromptTypes
{
    public const string RoomSetup = "ROOM_SETUP";
    public const string Mulligan = "MULLIGAN";
    public const string MainAction = "MAIN_ACTION";
    public const string StackPriority = "STACK_PRIORITY";
    public const string SpellDuelFocus = "SPELL_DUEL_FOCUS";
    public const string SpellDuelAction = "SPELL_DUEL_ACTION";
    public const string BattleDeclaration = "BATTLE_DECLARATION";
    public const string AssignCombatDamage = "ASSIGN_COMBAT_DAMAGE";
    public const string PayCost = "PAY_COST";
    public const string OrderTriggers = "ORDER_TRIGGERS";
    public const string TaskQueue = "TASK_QUEUE";
    public const string Wait = "WAIT";
    public const string MatchResult = "MATCH_RESULT";
}

public sealed record ActionPromptContractDto(
    string PromptKind,
    string CandidateAction,
    IReadOnlyList<string> RequiredPayload,
    IReadOnlyList<string> LegalChoices,
    IReadOnlyList<string> ValidationErrors,
    IReadOnlyList<string> VisibleMetadata,
    IReadOnlyList<string> HiddenMetadata);

public static class ActionPromptContracts
{
    public static ActionPromptContractDto PayCost { get; } = new(
        PromptTypes.PayCost,
        CommandTypes.PayCost,
        ["paymentId", "paymentWindow", "paymentChoiceIds"],
        ["candidate.metadata.paymentChoices", "candidate.metadata.paymentResourceChoices"],
        [ErrorCodes.InvalidPayload, ErrorCodes.PhaseNotAllowed, ErrorCodes.InsufficientCost],
        ["paymentId", "paymentWindow", "cost", "paymentChoices", "paymentResourceChoices"],
        ["serverPaymentState", "resourceLedgerBeforePayment"]);

    public static ActionPromptContractDto AssignCombatDamage { get; } = new(
        PromptTypes.AssignCombatDamage,
        CommandTypes.AssignCombatDamage,
        ["battleId", "battlefieldId", "assignments[].sourceObjectId", "assignments[].targetObjectId", "assignments[].damage"],
        ["candidate.metadata.assignmentChoices", "candidate.metadata.legalTargets", "candidate.metadata.battleParticipants"],
        [ErrorCodes.InvalidPayload, ErrorCodes.PhaseNotAllowed, ErrorCodes.InvalidTarget],
        ["battleId", "battlefieldId", "assigningPlayerId", "damagePool", "legalTargets", "existingDamage", "lethalDamageThreshold", "requiredAssignments", "assignmentChoices", "battleParticipants"],
        ["battleState", "participantControllerIds", "damageLedger"]);

    public static ActionPromptContractDto OrderTriggers { get; } = new(
        PromptTypes.OrderTriggers,
        CommandTypes.OrderTriggers,
        ["orderedTriggerIds"],
        ["candidate.metadata.triggerChoices", "candidate.metadata.orderedTriggerIds"],
        [ErrorCodes.InvalidPayload, ErrorCodes.PhaseNotAllowed, ErrorCodes.InvalidTarget],
        ["orderingPlayerId", "triggerIds", "triggers", "triggerChoices", "legalOrderingConstraints", "triggeredByEventKind"],
        ["triggerQueue"]);

    public static IReadOnlyDictionary<string, ActionPromptContractDto> ByPromptKind { get; } =
        new Dictionary<string, ActionPromptContractDto>(StringComparer.Ordinal)
        {
            [PayCost.PromptKind] = PayCost,
            [AssignCombatDamage.PromptKind] = AssignCombatDamage,
            [OrderTriggers.PromptKind] = OrderTriggers
        };
}

public sealed record PromptViewDto(
    string Type,
    string Title,
    string Message,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? RelatedBattlefieldId = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? RelatedStackItemId = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? RelatedBattleId = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? RelatedSpellDuelId = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] int? MinSelection = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] int? MaxSelection = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] IReadOnlyDictionary<string, object?>? Metadata = null);

public sealed record ActionPromptDto(
    string PlayerId,
    bool Actionable,
    string Reason,
    IReadOnlyList<string> Actions,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? PromptId = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] long? SnapshotTick = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] IReadOnlyList<ActionPromptCandidateDto>? Candidates = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] PromptViewDto? View = null);

public sealed record ActionPromptCandidateDto(
    string Action,
    string Label,
    bool Enabled,
    string Reason,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] IReadOnlyList<ActionPromptChoiceDto>? Sources = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] IReadOnlyList<ActionPromptChoiceDto>? Targets = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] IReadOnlyList<ActionPromptChoiceDto>? Destinations = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] IReadOnlyList<ActionPromptChoiceDto>? Modes = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] IReadOnlyList<ActionPromptChoiceDto>? OptionalCosts = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] IReadOnlyDictionary<string, object?>? Metadata = null);

public sealed record ActionPromptChoiceDto(
    string Id,
    string Label,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? Reason = null);

public sealed record ClientIntentDto(
    string RoomId,
    string PlayerId,
    string ClientIntentId,
    JsonElement Cmd);

File: src/Riftbound.DevUi/src/types/protocol.ts
export type ConnectionStatus = "idle" | "connecting" | "connected" | "reconnecting" | "resyncing" | "disconnected" | "error";

export type WsServerMessage<T = unknown> = {
  type: string;
  roomId: string;
  playerId: string;
  serverTick: number;
  payload: T;
  protocolVersion: number;
  schemaVersion: number;
};

export type PlayerSessionDto = {
  playerId: string;
  seat: string;
  reconnectToken: string;
};

export type ErrorDto = {
  code: string;
  message: string;
};

export type CommandReceiptDto = {
  accepted: boolean;
  clientIntentId: string;
  cmdType: string;
  errorCode?: string | null;
  message: string;
  playerId: string;
  promptId?: string | null;
  roomId: string;
  serverTick: number;
  snapshotTick?: number | null;
  state: "ACCEPTED" | "FAILED" | "REJECTED" | (string & {});
};

export type GameEventObjectRef = {
  objectId: string;
  role: string;
  cardNo?: string | null;
  ownerId?: string | null;
  controllerId?: string | null;
  zone?: string | null;
  battlefieldObjectId?: string | null;
  isFaceDown?: boolean;
  isHidden?: boolean;
};

export type GameEvent = {
  kind: string;
  description: string;
  objectRefs?: GameEventObjectRef[] | null;
  payload: Record<string, unknown>;
};

export type BattlefieldResolutionView = {
  resolutionId?: string;
  tick?: number;
  kind?: string;
  reason?: string;
  battlefieldObjectId?: string;
  playerId?: string | null;
  previousControllerId?: string | null;
  controllerId?: string | null;
  sourceObjectId?: string | null;
  participantObjectIds?: string[];
  relatedEventKinds?: string[];
};

export type BattleResolutionView = {
  resolutionId?: string;
  tick?: number;
  kind?: string;
  reason?: string;
  battlefieldId?: string;
  attackingPlayerId?: string | null;
  defendingPlayerId?: string | null;
  winnerPlayerId?: string | null;
  attackerObjectIds?: string[];
  defenderObjectIds?: string[];
  survivingAttackerObjectIds?: string[];
  survivingDefenderObjectIds?: string[];
  destroyedObjectIds?: string[];
  relatedEventKinds?: string[];
};

export type StackItemView = {
  stackItemId?: string;
  controllerId?: string;
  sourceObjectId?: string;
  effectKind?: string;
  cardNo?: string | null;
  targetObjectIds?: string[];
  damageAmount?: number;
  destination?: string;
};

export type PendingTaskView = {
  taskId?: string;
  kind?: string;
  status?: string;
  reason?: string;
  battlefieldObjectId?: string;
  participantControllerIds?: string[];
  participantObjectIds?: string[];
  actingPlayerId?: string | null;
  stackItemIds?: string[];
  spellDuelId?: string;
  battleId?: string;
};

export type PendingTaskQueueView = {
  activeTaskId?: string | null;
  hasTasks?: boolean;
  isBlocking?: boolean;
  phase?: string;
  tasks?: PendingTaskView[];
};

export type TriggerQueueItemView = {
  triggerId?: string;
  controllerId?: string;
  sourceObjectId?: string;
  sourceVisibility?: string;
  effectKind?: string;
  triggeredByEventKind?: string;
};

export type TurnWindowView = {
  state?: string;
  isSpellDuel?: boolean;
  isClosed?: boolean;
  hasStack?: boolean;
  actingPlayerId?: string | null;
};

export type RuleQueueCoverageSnapshotRow = {
  key?: "battle" | "payment" | "stack" | "trigger" | "window" | (string & {});
  liveCount?: number;
  evidenceKeys?: string[];
};

export type MatchTimingView = Record<string, unknown> & {
  battle?: { isActive?: boolean } | null;
  battleResolutions?: BattleResolutionView[];
  battlefieldResolutions?: BattlefieldResolutionView[];
  battlefieldTasks?: PendingTaskView[];
  pendingPayment?: Record<string, unknown> | null;
  pendingTaskQueue?: PendingTaskQueueView | null;
  phase?: string;
  ruleQueueCoverage?: RuleQueueCoverageSnapshotRow[] | null;
  roomStatus?: string;
  timingState?: string;
  triggerQueue?: TriggerQueueItemView[];
  turnWindow?: TurnWindowView | null;
};

export type ActionPromptChoiceDto = {
  id: string;
  label: string;
  objectIds?: string[] | null;
  reason?: string | null;
};

export type ActionPromptSelectionChoiceDto = {
  id: string;
  label: string;
  objectIds: string[];
  reason?: string | null;
};

export type ActionPromptSelectionStepDto = {
  role: "source" | "target" | "destination" | "mode" | "optionalCost" | (string & {});
  label: string;
  required: boolean;
  choices: ActionPromptSelectionChoiceDto[];
};

export type ActionPromptCommandTemplateBindingDto = {
  field: string;
  label?: string | null;
  source:
    | "selectedSource"
    | "selectedTarget"
    | "selectedTargets"
    | "selectedDestination"
    | "selectedMode"
    | "selectedOptionalCosts"
    | "candidateMetadata"
    | "requirementMetadata"
    | (string & {});
  required?: boolean;
  asArray?: boolean;
  omitEmpty?: boolean;
  metadataKey?: string | null;
  metadataKeys?: string[] | null;
};

export type ActionPromptCommandTemplateDto = {
  cmdType: string;
  bindings: ActionPromptCommandTemplateBindingDto[];
};

export type ActionPromptComposerDto = {
  supported: boolean;
  reason: string;
  selectionRoles: string[];
  requiredSelectionRoles: string[];
  commandFields: string[];
};

export type ActionPromptCandidateDto = {
  action: string;
  label: string;
  enabled: boolean;
  reason: string;
  sources?: ActionPromptChoiceDto[] | null;
  targets?: ActionPromptChoiceDto[] | null;
  destinations?: ActionPromptChoiceDto[] | null;
  modes?: ActionPromptChoiceDto[] | null;
  optionalCosts?: ActionPromptChoiceDto[] | null;
  metadata?: Record<string, unknown> | null;
  selectionSteps?: ActionPromptSelectionStepDto[] | null;
  commandTemplate?: ActionPromptCommandTemplateDto | null;
  composer?: ActionPromptComposerDto | null;
};

export type ActionPromptObjectCandidateDto = {
  action: string;
  label: string;
  enabled: boolean;
  reason: string;
  roles: string[];
  commandType?: string | null;
  requiredCommandFields?: string[] | null;
  commandFields?: string[] | null;
  composer?: ActionPromptComposerDto | null;
};

export type ActionPromptObjectInspectionRowDto = {
  key: string;
  label: string;
  value: string;
  tone?: string | null;
};

export type ActionPromptObjectInspectionGroupDto = {
  key: string;
  title: string;
  rows: ActionPromptObjectInspectionRowDto[];
  emptyLabel?: string | null;
};

export type ActionPromptObjectInspectionDto = {
  source: string;
  boundary: string;
  summaryRows: ActionPromptObjectInspectionRowDto[];
  groups: ActionPromptObjectInspectionGroupDto[];
};

export type ActionPromptInspectionRowDto = {
  key: string;
  label: string;
  value: string;
  tone?: string | null;
};

export type ActionPromptInspectionGroupDto = {
  key: string;
  title: string;
  rows: ActionPromptInspectionRowDto[];
  emptyLabel?: string | null;
};

export type ActionPromptInspectionDto = {
  source: string;
  boundary: string;
  summaryRows: ActionPromptInspectionRowDto[];
  groups: ActionPromptInspectionGroupDto[];
};

export type ActionPromptObjectContextDto = {
  objectId: string;
  enabledCandidateCount: number;
  disabledCandidateCount: number;
  candidates: ActionPromptObjectCandidateDto[];
  inspection?: ActionPromptObjectInspectionDto | null;
  source?: string | null;
  boundary?: string | null;
};

export type KnownPromptType =
  | "ROOM_SETUP"
  | "MULLIGAN"
  | "MAIN_ACTION"
  | "STACK_PRIORITY"
  | "SPELL_DUEL_FOCUS"
  | "SPELL_DUEL_ACTION"
  | "BATTLE_DECLARATION"
  | "HAND_CHOICE"
  | "ASSIGN_COMBAT_DAMAGE"
  | "PAY_COST"
  | "ORDER_TRIGGERS"
  | "TASK_QUEUE"
  | "WAIT"
  | "MATCH_RESULT";

export type PromptType = KnownPromptType | (string & {});

export type PromptViewDto = {
  type: PromptType;
  title: string;
  message: string;
  relatedBattlefieldId?: string | null;
  relatedStackItemId?: string | null;
  relatedBattleId?: string | null;
  relatedSpellDuelId?: string | null;
  minSelection?: number | null;
  maxSelection?: number | null;
  metadata?: Record<string, unknown> | null;
  responsibility?: PromptResponsibilityDto | null;
};

export type PromptResponsibilityDto = {
  promptType: PromptType;
  promptPlayerId: string;
  responsiblePlayerId?: string | null;
  isResponsiblePlayer: boolean;
  actionableForPromptPlayer: boolean;
  state: string;
  nextStep: string;
  queueCounts: Record<string, number>;
  relatedObjectIds: string[];
};

export type ActionPromptContractDto = {
  promptKind: string;
  candidateAction: string;
  requiredPayload: string[];
  legalChoices: string[];
  validationErrors: string[];
  visibleMetadata: string[];
  hiddenMetadata: string[];
};

export type ActionPromptContracts = Record<string, ActionPromptContractDto>;

export type ActionPromptDto = {
  playerId: string;
  actionable: boolean;
  reason: string;
  actions: string[];
  promptId?: string | null;
  snapshotTick?: number | null;
  candidates?: ActionPromptCandidateDto[] | null;
  contract?: ActionPromptContractDto | null;
  inspection?: ActionPromptInspectionDto | null;
  objectContexts?: ActionPromptObjectContextDto[] | null;
  view?: PromptViewDto | null;
};

export type RunePoolView = {
  mana?: number;
  power?: number;
  totalPower?: number;
  untypedPower?: number;
  powerByTrait?: Record<string, number>;
};

export type ZoneView = {
  mainDeckCount?: number;
  runeDeckCount?: number;
  hand?: string[];
  handHidden?: number;
  base?: string[];
  baseCards?: string[];
  baseRunes?: string[];
  battlefields?: string[];
  graveyard?: string[];
  banished?: string[];
  legendZone?: string[];
  championZone?: string[];
};

export type CardObjectView = {
  objectId?: string;
  cardNo?: string | null;
  damage?: number;
  basePower?: number;
  effectivePower?: number;
  power?: number;
  untilEndOfTurnPowerModifier?: number;
  isExhausted?: boolean;
  isFaceDown?: boolean;
  isAttacking?: boolean;
  isDefending?: boolean;
  tags?: string[];
  untilEndOfTurnEffects?: string[];
  manaCost?: number;
  attachedToObjectId?: string | null;
  ownerId?: string | null;
  controllerId?: string | null;
  location?: Record<string, unknown> | null;
};

export type PlayerSnapshotView = {
  id?: string;
  name?: string;
  seat?: string;
  ready?: boolean;
  deckSubmitted?: boolean;
  mulliganCompleted?: boolean;
  handSize?: number;
  score?: number;
  experience?: number;
  cardsPlayedThisTurn?: number;
  runePool?: RunePoolView;
  zones?: ZoneView;
  objects?: Record<string, CardObjectView>;
};

export type BattlefieldSnapshotView = {
  battlefieldObjectId?: string;
  zonePlayerId?: string;
  cardNo?: string | null;
  controllerId?: string | null;
  status?: string;
  contested?: boolean;
  occupantObjectIds?: string[];
  occupantControllerIds?: string[];
  unitsBySide?: Record<string, string[]>;
  standbyObjectIds?: string[];
  faceDownStandbyCount?: number;
  pendingTaskKinds?: string[];
  scoredThisTurn?: boolean;
  scoredPlayerId?: string | null;
  scoreStatus?: string | null;
};

export type SnapshotDto = {
  tick: number;
  turnNumber: number;
  activePlayerId: string;
  players: Record<string, PlayerSnapshotView>;
  lanes: Record<string, unknown>;
  stack: StackItemView[];
  timing: MatchTimingView;
  turnState: string;
};

export type SubmitDeckCommand = {
  cmdType: "SUBMIT_DECK";
  legendCardNo: string;
  championCardNo: string;
  mainDeck: string[];
  runeDeck: string[];
  battlefields: string[];
};

type PromptStampedCommand = {
  promptId?: string | null;
  snapshotTick?: number | null;
};

export type CombatDamageAssignmentDto = {
  sourceObjectId: string;
  targetObjectId: string;
  damage: number;
};

export type GameCommand = PromptStampedCommand & (
  | SubmitDeckCommand
  | { cmdType: "READY" }
  | { cmdType: "MULLIGAN"; handObjectIds: string[] }
  | { cmdType: "PASS_PRIORITY" }
  | { cmdType: "PASS_FOCUS" }
  | { cmdType: "PASS" }
  | { cmdType: "END_TURN" }
  | { cmdType: "SURRENDER" }
  | { cmdType: "PLAY_CARD"; sourceObjectId: string; cardNo: string; targetObjectIds: string[]; mode?: string; optionalCosts?: string[]; destination?: string }
  | { cmdType: "HIDE_CARD"; sourceObjectId: string; cardNo: string; destination?: string; optionalCosts?: string[] }
  | { cmdType: "REVEAL_CARD"; sourceObjectId: string; cardNo: string; mode?: string; destination?: string; optionalCosts?: string[]; targetObjectIds?: string[] }
  | { cmdType: "TAP_RUNE"; sourceObjectId: string }
  | { cmdType: "RECYCLE_RUNE"; sourceObjectId: string }
  | { cmdType: "MOVE_UNIT"; sourceObjectId: string; origin?: string; destination?: string; optionalCosts?: string[] }
  | { cmdType: "ASSEMBLE_EQUIPMENT"; sourceObjectId: string; targetObjectId?: string; optionalCosts?: string[] }
  | { cmdType: "DECLARE_BATTLE"; battlefieldId?: string; attackerObjectIds?: string[]; defenderObjectIds?: string[]; battlefieldTargetObjectIds?: string[]; optionalCosts?: string[] }
  | { cmdType: "ACTIVATE_ABILITY"; sourceObjectId: string; abilityId: string; targetObjectIds: string[]; optionalCosts?: string[] }
  | { cmdType: "LEGEND_ACT"; sourceObjectId: string; abilityId: string; targetObjectIds: string[]; optionalCosts?: string[] }
  | { cmdType: "PAY_COST"; paymentId?: string; paymentWindow?: string; paymentChoiceIds?: string[] | null }
  | { cmdType: "ASSIGN_COMBAT_DAMAGE"; battleId?: string; battlefieldId?: string; assignments?: CombatDamageAssignmentDto[] | null }
  | { cmdType: "ORDER_TRIGGERS"; orderedTriggerIds?: string[] | null; triggerIds?: string[] | null }
  | { cmdType: "CHOOSE_HAND_CARDS"; choiceId: string; choiceWindow: string; chosenObjectIds: string[] }
);
